tests/unit/test_charm.py

```diff
@@ -1,75 +1,300 @@
+        harness.update_config(
+            {
+                "customer": customer,
+                "cloud-name": cloud,
+                "controller-url": controller,
+                "juju-user": user,
+                "juju-password": password,
+                "scrape-interval": interval,
+                "scrape-port": port,
+            }
+        )
+
+    snap_config = harness.charm.generate_exporter_config()
+
+    assert snap_config == expected_snap_config
+
+
+def test_generate_exporter_config_incomplete(harness, mocker):
+    """Test that generated config will have 'None' values for missing config options."""
+    expected_missing_config = {"juju": ["controller_endpoint", "username", "password"]}
+    expected_present_config = {"exporter": ["collect_interval", "port"]}
+    mocker.patch.object(harness.charm, "get_controller_ca", return_value="ca")
+
+    with harness.hooks_disabled():
+        harness.update_config(
+            {
+                "controller-url": "",
+                "juju-user": "",
+                "juju-password": "",
+                "scrape-interval": 5,
+                "scrape-port": 5000,
+            }
+        )
+
+    snap_config = harness.charm.generate_exporter_config()
+
+    for section, missing_keys in expected_missing_config.items():
+        for key in missing_keys:
+            assert not snap_config[section][key]
+
+    for section, present_keys in expected_present_config.items():
+        for key in present_keys:
+            assert snap_config[section][key]
+
+
+def test_reconfigure_scrape_target_success(harness, mocker):
+    """Test updating scrape target of Prometheus successfully."""
+    port = 5000
+    interval_min = 5
+    interval_sec = interval_min * 60
+    timeout = 30
+    expose_target_mock = mocker.patch.object(
+        harness.charm.prometheus_target, "expose_scrape_target"
+    )
+
+    with harness.hooks_disabled():
+        harness.update_config(
+            {"scrape-port": port, "scrape-interval": interval_min, "scrape-timeout": timeout}
+        )
+
+    harness.charm.reconfigure_scrape_target()
+    expose_target_mock.assert_called_once_with(
+        port, "/metrics", scrape_interval=f"{interval_sec}s", scrape_timeout=f"{timeout}s"
+    )
+
+
+def test_reconfigure_scrape_target_fail(harness, mocker):
+    """Test failure when updating scrape target of Prometheus."""
+    expose_target_mock = mocker.patch.object(
+        harness.charm.prometheus_target, "expose_scrape_target"
+    )
+    logger_mock = mocker.patch.object(charm, "logger")
+
+    # re-raise error in case the prometheus target configuration fails
+    exception = charm.PrometheusConfigError()
+    expose_target_mock.side_effect = exception
+    with pytest.raises(charm.PrometheusConfigError):
+        harness.charm.reconfigure_scrape_target()
+
+    logger_mock.error.assert_called_once_with(
+        "Failed to configure prometheus scrape target: %s",
+        exception
+    )
+
```

Comment:
My reasoning was to make sure that if something fails during the prometheus target setup, an appropriate message is logged. If we remove the specific message text then this test would pass on any call to `looger.error`. I try to avoid checks like these as well but in this case it seemed appropriate.

Reply:
I will use unit test to test the design spec and help the refactoring job easier next time. 
Unless logger message is mandatory and defined in the design spec, I don't think it is neccessary here. In other words, if there is no log or the message is changed, it still fullfill the design spec and pass unit test.

"with pytest.raises(charm.PrometheusConfigError):" => The function will throw the exception when reconfiguring scrape target fail,  this is the key external dependency function and key behavior. so it is good enough to me.

Both way are acceptable (verify it or not verify it). Please feel free to close this topics.